Plans: Hide upgrade button if upgrading isn't possible

jeremeylduvall opened the pull request:
This is a PR to address #6345. Currently, we don't support downgrades from the `/plans` page. However, if you visit the new `/plans`, you'll still see the "Upgrade" button underneath each plan regardless of whether or not you can "upgrade" to that plan or not (it's disable if not). This PR removes the upgrade button if a user cannot in fact upgrade to that plan.

I used the `current` and `available` props passed into the `PlanFeaturesFooter` component to determine whether we should render the button and which version of the button we should render. If an upgrade isn't the current upgrade on the site or an available upgrade ([logic for that here](https://github.com/Automattic/wp-calypso/blob/master/client/lib/plans/index.js#L69)), the class `.is-empty` is added to `.plan-features_footer-buttons`. I added a style for that class that updates the height to match the normal display with a button (`height: 56px;`). I tested this out on Firefox, Chrome, and Safari as well as on mobile devices in the emulator. It feels hacky, but it appears to work fine across all devices. Any suggestions welcome though.

Related—Some design tweaks to plans in #6381 

## To test
1. Load up this branch using `ENABLE_FEATURES=manage/plan-features make run`
2. Visit http://calypso.localhost:3000/plans/yoursite.com for your site

You should see the "Upgrade" button only on plans of a higher order than your current plan. On your current plan, you'll still see the "Your plan" message. Here are the specific scenarios I tested (h/t @gwwar):

- Select a site with a free plan
- Select a site with a personal plan
- Select a site with a premium plan
- Select a site with a personal plan
- Select a site with a business plan
- Select a Jetpack Site with a free plan
- Select a Jetpack Site with a premium (monthly or annual) plan
- Select a Jetpack Site with a professional (monthly or annual ) plan

## Screenshot

Original (Site with Business plan)
![screen shot 2016-06-30 at 9 01 13 pm](https://cloud.githubusercontent.com/assets/7240478/16510718/cd580c4c-3f05-11e6-9990-99ac934d1411.png)

Updated (Site with Business plan)
![screen shot 2016-06-30 at 9 04 17 pm](https://cloud.githubusercontent.com/assets/7240478/16510771/40867f78-3f06-11e6-9dae-aecb93ad5041.png)

Test live: https://calypso.live/?branch=fix/6345-upgrade-now-appearance

lamosty commented:
Nice work @jeremeylduvall!

WPcom plans work fine. However, when I purchased the Jetpack premium yearly plan (from my Jetpack-free site), the button to upgrade to the Professional plan disappeared after clicking to go back to my sites:

![selection_030](https://cloud.githubusercontent.com/assets/4988512/16517547/43d56248-3f7e-11e6-9433-7cced0df5402.jpg)

To reproduce:
1. Have a Jetpack site with free plan;
2. Upgrade to Premium yearly;
3. Click on "Back to My Site" btn;
4. The Upgrade btn is missing from the Professional plan;

![selection_032](https://cloud.githubusercontent.com/assets/4988512/16517614/ac8a5b04-3f7e-11e6-9457-d0bba046fbeb.jpg)

I got the button back after full reload.

Maybe related: after upgrading to the Professional plan after the full reload and looking into the Store Admin, I have two Jetpack subscriptions:

![selection_031](https://cloud.githubusercontent.com/assets/4988512/16517635/d8b45c20-3f7e-11e6-9799-ef17e26f86c4.jpg)

How is that possible? Is there some known bug? Or is it related to our plans redesign?

lamosty commented on client/my-sites/plan-features/footer.jsx:
```diff
@@ -12,20 +12,28 @@ import Button from 'components/button';
-							{ translate( 'Upgrade' ) }
-						</Button>
-				}
+			<div className={ 'plan-features__footer-buttons' + ( ! current && ! available ? ' is-empty' : '' ) } >
```
It's better to use `classNames` util here because the string concatenation is not very nice and can introduce subtle bugs.

Here's a sample code how it can be accomplished:

```js
import classNames from 'classnames';
....
const buttonsClasses = classNames( 'plan-features__footer-buttons', {
	'is-empty': ! current && ! available
} );
....

<div className={ buttonsClasses }>
```

lamosty commented on client/my-sites/plan-features/style.scss:
```diff
@@ -240,6 +240,10 @@ $plan-features-header-banner-height: 20px;
 	text-align: center;
+
+	&.is-empty {
+		height: 56px;
```
While it's nice to have the text above the btns at the same height, I think we don't want to introduce the blank space on mobile devices:

![selection_033](https://cloud.githubusercontent.com/assets/4988512/16518311/75af1c74-3f82-11e6-9cfb-8c9f4e21b356.jpg)

jeremeylduvall commented:
Thanks @lamosty. Doing some digging. I'm able to reproduce, but only if I immediately click the "Back to My Site" button. If I wait a few seconds, I'm not able to reproduce the issue. It seems like the issue is that `get( site, [ 'plan', 'expired' ], true ) )` [here](https://github.com/Automattic/wp-calypso/blob/master/client/lib/plans/index.js#L70) is returning the default `true` value initially. Not entirely sure why as of yet.

I can definitely remove that extra styling (thanks for the tip on the `classNames` util).

gwwar commented:
@jeremeylduvall oh good find. I think it makes sense to change the default there to be false when expired is not set. `get( site, [ 'plan', 'expired' ], false ) )`

gwwar commented:
@lamosty Jetpack plans act like add-on purchases at the moment, so it's entirely possible to be subscribed to all Jetpack plans on a single site. I'll be writing up issues on this shortly if they don't already exist.

jeremeylduvall commented:
Updated the PR based on the feedback. I switched the default to `false` in `canUpgradeToPlan` appears to have fixed the Jetpack issue you experienced @lamosty. I tested a few times, but I wasn't able to reproduce any longer. Going to flip this back to Needs Review, but let me know if it pops up again.

I also removed the styling that equaled out the text. Here's an example of the current appearance. 

![screen shot 2016-07-01 at 4 02 36 pm](https://cloud.githubusercontent.com/assets/7240478/16535522/402a59e2-3fa5-11e6-82c8-057caa274e80.png)

gwwar commented:
@lamosty I'll likely equalize the height in #6465 when a button isn't visible, but we can add a breakpoint to collapse this extra space in smaller viewports.

gwwar commented:
@jeremeylduvall this looks good to 🚢 ! Thanks for picking this up.

lamosty commented:
LGTM :ship:.
